content/Building an Engine-modularized API in Rails - Guillermo Leopold.md

```diff
@@ -0,0 +1,260 @@
+  * When installing a new gem, in order to be able to use it across the whole engine, we should require it in lib/engine_name.rb
+  * As a result of the engine namespace models, services, controllers and even the tables in the database will be namespaced too.
+  * Some tips I suggest:
+    * Create an engine that is visible in all the engines across the application. This engine will help us to reduce repeated code and may contain things that we want to use in all the engines. In our case, we have an engine called Api that handles all the requests validations (checking headers, render some common errors, etc) and all the main controller from other engines inherit from a controller located in the Api engine
+    * Create “libraries” that work as interfaces between our API and external services. I.e: In our case we have a gem called Catalog Providers and inside it we have services for each provider that we use in our system (like Target, Amazon, BestBuy)
+    * It might be useful to create a few scripts to run migrations, run rspec or bundle install across all the engines. The scripts will iterate over the engines folder and run the command inside each one.
+  * Finally, it's worth to mention two ways to handle the escenario where we need to trigger an action in an engine where we don't have access from the engine we are currently working:
+     * Using an async pattern like queue with pub/sub where the element that is running publishes a message in the queue and the element that is responsible of running the desired triggered action is subscribed to that queue and reads any message to know which action has to take.
+     * The other pattern that works fine is implementing an observer pattern, in our application we followed this pattern as Rails already has a good implementation of it for ActiveRecord. This pattern works by having classes called observers that are responsible of beware of any action happening in the "observed class" and trigger actions depending on what action was taken in the followed class.
+
+Using the **User** and **Payment** engines where **Payment** can access to **User** but **User** can't access to **Payment** to avoid having a circular dependency, let's see step by step how we implement the **observer pattern** in our application:
+
+1. Create an observer class inside our models folder in the **Payment engine** as we want to trigger an action in this engine as a consequence of an action in the **User** engine.
+
+2. Set up the observer in the **engine.rb** file by adding the following line:
+ ```Ruby
+ config.active_record.observers = :'payment/user_customer_observer'
+ ```
+
+3. Implement your observer, we need to specify in which model we want to observe and implement an action according that behave similar to a callback in ActiveRecord. It might look like this:
```

Comment:
```suggestion
3. Implement your observer, we need to specify which model we want to observe and implement an action according that behaves similar to a callback in ActiveRecord. It might look like this:
```